lib/gtl_formatter.rb

```diff
@@ -41,17 +88,102 @@ def self.format_cols(view, row, controller)
+      elsif COLUMN_WITH_IMAGE.keys.include?(col)
+        # Generate html for the list icon
+        icon, icon2, image = send(COLUMN_WITH_IMAGE[col], record)
+        text = format_col_for_display(view, row, col)
+        item = {:title => text,
+                :image => ActionController::Base.helpers.image_path(image.to_s),
+                :icon  => icon,
+                :icon2 => icon2,
+                :text  => text}.compact
+      elsif COLUMN_WITH_ICON.keys.include?(col)
+        # Generate html for the list icon
+        icon = send(COLUMN_WITH_ICON[col], record)
+        text = format_col_for_display(view, row, col)
+        item = {:title => text,
+                :icon  => icon,
+                :text => text}.compact
+      elsif COLUMN_WITH_TIME.include?(col)
+        celltext = format_time_for_display(row, col)
       else
         celltext = format_col_for_display(view, row, col)
       end
 
-      item = {:text => celltext}
+      item ||= {:text => celltext}
       item[:span] = span if span.present?
       cols.push(item)
     end
+
+    # Append a button if @row_button is set and the button is defined in the related decorator
+    button = record.decorate.try(:gtl_button_cell) if options[:row_button]
+    cols.push(button) if button
+
     cols
   end
 
+  def self.format_icon_column(view, row, clickable)
+    item = listicon_item(view, row['id'])
+    icon, icon2, image = fonticon_or_fileicon(item)
+
+    # Clickable should be false only when it's explicitly set to false
+    {:title => clickable == false ? nil : _('View this item'),
+     :image => ActionController::Base.helpers.image_path(image.to_s),
+     :icon  => icon,
+     :icon2 => icon2}.compact
+  end
+
+  def self.listicon_item(view, id = nil)
+    id = @id if id.nil?
+
+    if @targets_hash
+      @targets_hash[id] # Get the record from the view
+    else
+      klass = view.db.constantize
+      klass.find(id)    # Read the record from the db
+    end
+  end
+
+  def self.fonticon_or_fileicon(item)
+    return nil unless item
+    decorated = item.decorate
+    [
+      decorated.try(:fonticon),
+      decorated.try(:secondary_icon),
+      decorated.try(:fileicon)
+    ]
+  end
+
+  def self.last_compliance_status_image(item)
+    case item.last_compliance_status
+    when true
+      "pficon pficon-ok"
+    when false
+      "pficon pficon-error-circle-o"
+    end
+  end
+
+  def self.authentication_status_image(item)
+    case item.authentication_status.downcase
+    when "error", "invalid"
+      "pficon pficon-error-circle-o"
+    when "valid"
+      "pficon pficon-ok"
+    when "none"
+      "pficon pficon-unknown"
+    else
+      "pficon pficon-warning-triangle-o"
+    end
+  end
+
+  def self.normalized_state_image(item)
+    NORMALIZED_STATE_ICON[item.normalized_state]
+  end
+
+  def self.has_policies_image(item)
+    item.has_policies ? "fa fa-shield" : "pficon pficon-pending-2"
```

Comment:
And this one is causing the tag set of SQL queries.

Again, I think this can be solved by  using `row['has_policies']` instead of `record.has_policies`

Reply:
I will be removing this bit of code, @chessbyte and i have decided to remove `Protected` column from list view and add it on a summary screen instead.